File: backend/Backend/src/main/java/com/sgw/backend/entity/Advertisement.java
package com.sgw.backend.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@EqualsAndHashCode(callSuper = true)
@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor

public class Advertisement extends Billable {

//    @ElementCollection
//    private List<String> images;

    private String name;

    private String image;

    private String description;

    private LocalDate startDate;

    private LocalDate endDate;

    private String adCreatorUsername;

    // TODO: BILLABLE PRICE IS THE ADVERTISEMENT BUDGET

    // cost per view is set by owner to decide how much he want to pay per impression
    private BigDecimal costPerImpression;

    // this allows the us or users to see the number of max impression they can have left
    // TODO: HANDLE AND CALCULATE
    private Long impressionsLeft;

    // KIV THIS FEATURE
//    private BigDecimal costPerClick;

    // TODO: HANDLE AND CALCULATE
    private BigDecimal budgetLeft;

    // TODO: Reach and Impression IDS will store the ID of who viewed the add

    //impression total number of display not unique
    // TODO: HANDLE AND CALCULATE
    @ElementCollection
    private List<Long> impressionIds;

    private Long impressionCount;

    //reach is the total number of unique people who see your content
    // TODO: HANDLE AND CALCULATE
    @ElementCollection
    private Map<Long, Long> reachIds;

    private Long reachCount;

    private AdvertisementStatusEnum advertisementStatus;

    private String rejectionReason;

    @JsonIgnore
    @ManyToOne(optional = false)
    private Owner owner;


    @Override
    public String getLabel() {
        return name;
    }
//    public Advertisement() {
//        this.impressionIds = new ArrayList<>();
//        this.reachIds = new HashMap<>();
//    }
//
//    public List<String> getImages() {
//        return images;
//    }
//
//    public void setImages(List<String> images) {
//        this.images = images;
//    }
//
//    public String getDescription() {
//        return description;
//    }
//
//    public void setDescription(String description) {
//        this.description = description;
//    }
//
//    public LocalDate getStartDate() {
//        return startDate;
//    }
//
//    public void setStartDate(LocalDate startDate) {
//        this.startDate = startDate;
//    }
//
//    public LocalDate getEndDate() {
//        return endDate;
//    }
//
//    public void setEndDate(LocalDate endDate) {
//        this.endDate = endDate;
//    }
//
//    public String getRequestorEmail() {
//        return requestorEmail;
//    }
//
//    public void setRequestorEmail(String requestorEmail) {
//        this.requestorEmail = requestorEmail;
//    }
//
//    public String getRequestorName() {
//        return requestorName;
//    }
//
//    public void setRequestorName(String requestorName) {
//        this.requestorName = requestorName;
//    }
//
//    public String getRequestorPhoneNumber() {
//        return requestorPhoneNumber;
//    }
//
//    public void setRequestorPhoneNumber(String requestorPhoneNumber) {
//        this.requestorPhoneNumber = requestorPhoneNumber;
//    }
//
//    public BigDecimal getAmountCharge() {
//        return amountCharge;
//    }
//
//    public void setAmountCharge(BigDecimal amountCharge) {
//        this.amountCharge = amountCharge;
//    }
//
//    public AdvertisementStatusEnum getAdvertisementStatus() {
//        return advertisementStatus;
//    }
//
//    public void setAdvertisementStatus(AdvertisementStatusEnum advertisementStatus) {
//        this.advertisementStatus = advertisementStatus;
//    }
//
//    public Owner getOwner() {
//        return owner;
//    }
//
//    public void setOwner(Owner owner) {
//        this.owner = owner;
//    }
}

File: student_frontend/lib/model/advertisement.dart
class Advertisement {
  final int billableId;
  final double billablePrice;
  final String name;
  final String image;
  final String description;
  final String startDate;
  final String endDate;
  final String adCreatorUsername;
  final double costPerImpression;
  final int impressionsLeft;
  final double budgetLeft;
  final List<int> impressionIds;
  final int impressionCount;
  final Map<String, int> reachIds;
  final int reachCount;
  final String advertisementStatus;
  final String rejectionReason;

  Advertisement({
    required this.billableId,
    required this.billablePrice,
    required this.name,
    required this.image,
    required this.description,
    required this.startDate,
    required this.endDate,
    required this.adCreatorUsername,
    required this.costPerImpression,
    required this.impressionsLeft,
    required this.budgetLeft,
    required this.impressionIds,
    required this.impressionCount,
    required this.reachIds,
    required this.reachCount,
    required this.advertisementStatus,
    required this.rejectionReason,
  });

  factory Advertisement.fromJson(Map<String, dynamic> json) {
    return Advertisement(
      billableId: json['billableId'],
      billablePrice: json['billablePrice'],
      name: json['name'],
      image: json['image'],
      description: json['description'],
      startDate: json['startDate'],
      endDate: json['endDate'],
      adCreatorUsername: json['adCreatorUsername'],
      costPerImpression: json['costPerImpression'],
      impressionsLeft: json['impressionsLeft'],
      budgetLeft: json['budgetLeft'],
      impressionIds: List<int>.from(json['impressionIds']),
      impressionCount: json['impressionCount'],
      reachIds: Map<String, int>.from(json['reachIds']),
      reachCount: json['reachCount'],
      advertisementStatus: json['advertisementStatus'],
      rejectionReason: json['rejectionReason'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'billableId': billableId,
      'billablePrice': billablePrice,
      'name': name,
      'image': image,
      'description': description,
      'startDate': startDate,
      'endDate': endDate,
      'adCreatorUsername': adCreatorUsername,
      'costPerImpression': costPerImpression,
      'impressionsLeft': impressionsLeft,
      'budgetLeft': budgetLeft,
      'impressionIds': impressionIds,
      'impressionCount': impressionCount,
      'reachIds': reachIds,
      'reachCount': reachCount,
      'advertisementStatus': advertisementStatus,
      'rejectionReason': rejectionReason,
    };
  }

  @override
  String toString() {
    return '''
    Advertisement(
      billableId: $billableId,
      billablePrice: $billablePrice,
      name: $name,
      image: $image,
      description: $description,
      startDate: $startDate,
      endDate: $endDate,
      adCreatorUsername: $adCreatorUsername,
      costPerImpression: $costPerImpression,
      impressionsLeft: $impressionsLeft,
      budgetLeft: $budgetLeft,
      impressionIds: ${impressionIds.toString()},
      impressionCount: $impressionCount,
      reachIds: ${reachIds.toString()},
      reachCount: $reachCount,
      advertisementStatus: $advertisementStatus,
      rejectionReason: $rejectionReason
    )
  ''';
  }
}
